File: HW9/HW9.py
class TreeNode(object):

     def __init__(self, val=0, left=None, right=None):

         self.val = val

         self.left = left

         self.right = right

class HW9:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        if root is None:
            return None
        if root == p or root == q:
            return root
        if p.val < root.val and q.val< root.val :
            return self.lowestCommonAncestor(root.left, p, q)
        if p.val> root.val and q.val> root.val :
            return self.lowestCommonAncestor(root.right, p, q)
        else:
            return root
    
#Test Case 1   
root = TreeNode(4)
node3 = TreeNode(3)
node1 = TreeNode(1)
node8 = TreeNode(8)
node5 = TreeNode(5)
node9 = TreeNode(9)
    
root.left = node3
root.right = node8
node3.left = node1
node8.left = node5
node8.right = node9
    
hw9solution = HW9()
result = hw9solution.lowestCommonAncestor(root, node1, node5)
print("LCA:", result.val)
    
    
    

File: HW9/HW9.java
package HW9;

public class HW9 {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        //not necessary because p and q do exist in the tree so root cannot be null
        /*if(root == null){
            return null;
        }*/

        //we need a base case to return the root if either p or q are the root
        if (root == p|| root==q){
            //if p or q is root, then LCA has to be the root
            return root;
        }

        //if the p and q values are both smaller then the root then we go to the left and continue searching
        if(p.val<root.val && q.val<root.val){
        return lowestCommonAncestor(root.left, p, q);
        }
        //if p and q values are both bigger than the root then we go right because that is greater than the root
        if(p.val>root.val && q.val>root.val){
        return lowestCommonAncestor(root.right,p,q);
        }

        //if neither of the if statements pass, then we know that p and q might be on opposite sides of the root,
        //meaning that the only LCA will be the root, and so since p and q exist it must return the root
        else{
            return root;
        }

    }

    
    
    public static void main(String[] args)
    {
        //Test Case
        TreeNode root = new TreeNode(4);
        TreeNode node3 = new TreeNode(3); 
        TreeNode node1 = new TreeNode(1); 
        TreeNode node8 = new TreeNode(8);
        TreeNode node5 = new TreeNode(5);
        TreeNode node9 = new TreeNode(9);

        root.left = node3;
        root.right = node8;
        node3.left = node1;
        node8.left = node5;
        node8.right = node9;

        HW9 hw9 = new HW9();
        TreeNode result = hw9.lowestCommonAncestor(root, node1, node3);
         System.out.println("LCA: " + result.val);
     }
 }
